File: skills/distributed-systems/consensus-raft/resources/examples/python/raft_client.py
#!/usr/bin/env python3
"""
Production-ready etcd RAFT client examples

Demonstrates common patterns for working with etcd consensus:
- Basic key-value operations
- Transactions (atomic compare-and-swap)
- Leases (TTL-based keys for distributed locking)
- Watches (real-time change notifications)
- Leader election
- Configuration management

Requirements:
    pip install etcd3-py
"""

import time
import json
import threading
from typing import Optional, Callable
from contextlib import contextmanager

try:
    import etcd3
except ImportError:
    print("Error: etcd3-py not installed")
    print("Install with: pip install etcd3-py")
    exit(1)


class RAFTClient:
    """Production-ready etcd client with best practices"""

    def __init__(self, host='localhost', port=2379, timeout=10):
        """
        Initialize etcd client

        Args:
            host: etcd server hostname
            port: etcd server port
            timeout: Connection timeout in seconds
        """
        self.client = etcd3.client(host=host, port=port, timeout=timeout)
        self.host = host
        self.port = port

    def put(self, key: str, value: str, lease: Optional[int] = None) -> bool:
        """
        Write key-value pair (goes through RAFT consensus)

        Args:
            key: Key to write
            value: Value to write
            lease: Optional lease ID for TTL

        Returns:
            True if successful
        """
        try:
            if lease:
                self.client.put(key, value, lease=lease)
            else:
                self.client.put(key, value)
            return True
        except Exception as e:
            print(f"PUT error: {e}")
            return False

    def get(self, key: str) -> Optional[str]:
        """
        Read value for key (linearizable read from leader by default)

        Args:
            key: Key to read

        Returns:
            Value if found, None otherwise
        """
        try:
            value, metadata = self.client.get(key)
            return value.decode('utf-8') if value else None
        except Exception as e:
            print(f"GET error: {e}")
            return None

    def get_prefix(self, prefix: str) -> dict:
        """
        Get all keys with given prefix

        Args:
            prefix: Key prefix

        Returns:
            Dictionary of key-value pairs
        """
        try:
            results = {}
            for value, metadata in self.client.get_prefix(prefix):
                key = metadata.key.decode('utf-8')
                val = value.decode('utf-8')
                results[key] = val
            return results
        except Exception as e:
            print(f"GET_PREFIX error: {e}")
            return {}

    def delete(self, key: str) -> bool:
        """
        Delete key

        Args:
            key: Key to delete

        Returns:
            True if deleted
        """
        try:
            self.client.delete(key)
            return True
        except Exception as e:
            print(f"DELETE error: {e}")
            return False

    def compare_and_swap(self, key: str, expected_value: str, new_value: str) -> bool:
        """
        Atomic compare-and-swap using transaction

        Args:
            key: Key to update
            expected_value: Expected current value
            new_value: New value to set

        Returns:
            True if swap succeeded
        """
        try:
            success, responses = self.client.transaction(
                compare=[
                    self.client.transactions.value(key) == expected_value.encode('utf-8')
                ],
                success=[
                    self.client.transactions.put(key, new_value)
                ],
                failure=[
                    self.client.transactions.get(key)
                ]
            )
            return success
        except Exception as e:
            print(f"CAS error: {e}")
            return False

    def increment_counter(self, key: str) -> Optional[int]:
        """
        Atomically increment counter using transaction

        Args:
            key: Counter key

        Returns:
            New counter value, None on error
        """
        try:
            while True:
                current_value = self.get(key)
                current = int(current_value) if current_value else 0
                new_value = current + 1

                if self.compare_and_swap(key, str(current), str(new_value)):
                    return new_value

                # CAS failed, retry
                time.sleep(0.001)
        except Exception as e:
            print(f"INCREMENT error: {e}")
            return None

    @contextmanager
    def lease(self, ttl: int):
        """
        Context manager for lease (TTL-based key)

        Args:
            ttl: Time-to-live in seconds

        Usage:
            with client.lease(60) as lease_id:
                client.put('/ephemeral/key', 'value', lease=lease_id)
        """
        lease_obj = self.client.lease(ttl)
        try:
            yield lease_obj.id
        finally:
            lease_obj.revoke()

    def watch(self, key: str, callback: Callable, prefix: bool = False):
        """
        Watch for changes to key

        Args:
            key: Key to watch
            callback: Function to call on changes (receives event)
            prefix: If True, watch all keys with prefix

        Usage:
            def on_change(event):
                print(f"Changed: {event.key} = {event.value}")

            client.watch('/config/', on_change, prefix=True)
        """
        try:
            watch_id = self.client.add_watch_callback(
                key,
                callback,
                range_end=self.client.range_end_for_prefix(key) if prefix else None
            )
            return watch_id
        except Exception as e:
            print(f"WATCH error: {e}")
            return None


class DistributedLock:
    """Distributed lock using etcd leases"""

    def __init__(self, client: RAFTClient, lock_name: str, ttl: int = 60):
        """
        Initialize distributed lock

        Args:
            client: RAFT client
            lock_name: Unique lock identifier
            ttl: Lock TTL in seconds
        """
        self.client = client
        self.lock_key = f'/locks/{lock_name}'
        self.ttl = ttl
        self.lease_id = None
        self.keep_alive_thread = None
        self.stop_keep_alive = False

    def acquire(self, timeout: int = 10) -> bool:
        """
        Acquire lock (blocking with timeout)

        Args:
            timeout: Maximum time to wait in seconds

        Returns:
            True if acquired
        """
        start_time = time.time()

        while time.time() - start_time < timeout:
            # Create lease
            lease = self.client.client.lease(self.ttl)
            self.lease_id = lease.id

            # Try to acquire lock (create key if doesn't exist)
            success, _ = self.client.client.transaction(
                compare=[
                    self.client.client.transactions.version(self.lock_key) == 0
                ],
                success=[
                    self.client.client.transactions.put(
                        self.lock_key,
                        'locked',
                        lease=self.lease_id
                    )
                ],
                failure=[]
            )

            if success:
                # Start keep-alive thread
                self._start_keep_alive(lease)
                return True

            # Lock held by someone else, wait and retry
            time.sleep(0.1)

        return False

    def release(self):
        """Release lock"""
        self.stop_keep_alive = True
        if self.keep_alive_thread:
            self.keep_alive_thread.join()

        if self.lease_id:
            try:
                self.client.client.revoke_lease(self.lease_id)
            except:
                pass
            self.lease_id = None

    def _start_keep_alive(self, lease):
        """Start background thread to keep lease alive"""
        def keep_alive():
            while not self.stop_keep_alive:
                try:
                    lease.refresh()
                except:
                    break
                time.sleep(self.ttl / 3)  # Refresh at 1/3 TTL

        self.keep_alive_thread = threading.Thread(target=keep_alive, daemon=True)
        self.keep_alive_thread.start()

    def __enter__(self):
        """Context manager entry"""
        if not self.acquire():
            raise RuntimeError("Failed to acquire lock")
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit"""
        self.release()


class LeaderElection:
    """Leader election using etcd"""

    def __init__(self, client: RAFTClient, election_name: str, node_id: str, ttl: int = 60):
        """
        Initialize leader election

        Args:
            client: RAFT client
            election_name: Unique election identifier
            node_id: This node's identifier
            ttl: Leadership TTL in seconds
        """
        self.client = client
        self.election_key = f'/elections/{election_name}/leader'
        self.node_id = node_id
        self.ttl = ttl
        self.is_leader = False
        self.lease_id = None
        self.keep_alive_thread = None
        self.stop_keep_alive = False

    def campaign(self) -> bool:
        """
        Attempt to become leader

        Returns:
            True if became leader
        """
        # Create lease
        lease = self.client.client.lease(self.ttl)
        self.lease_id = lease.id

        # Try to become leader
        success, _ = self.client.client.transaction(
            compare=[
                self.client.client.transactions.version(self.election_key) == 0
            ],
            success=[
                self.client.client.transactions.put(
                    self.election_key,
                    self.node_id,
                    lease=self.lease_id
                )
            ],
            failure=[]
        )

        if success:
            self.is_leader = True
            self._start_keep_alive(lease)
            return True

        return False

    def resign(self):
        """Resign from leadership"""
        self.stop_keep_alive = True
        if self.keep_alive_thread:
            self.keep_alive_thread.join()

        if self.lease_id:
            try:
                self.client.client.revoke_lease(self.lease_id)
            except:
                pass
            self.lease_id = None

        self.is_leader = False

    def get_leader(self) -> Optional[str]:
        """
        Get current leader

        Returns:
            Leader node ID, None if no leader
        """
        leader = self.client.get(self.election_key)
        return leader

    def _start_keep_alive(self, lease):
        """Start background thread to keep lease alive"""
        def keep_alive():
            while not self.stop_keep_alive:
                try:
                    lease.refresh()
                except:
                    self.is_leader = False
                    break
                time.sleep(self.ttl / 3)

        self.keep_alive_thread = threading.Thread(target=keep_alive, daemon=True)
        self.keep_alive_thread.start()


# Example usage
if __name__ == '__main__':
    print("=== RAFT Client Examples ===\n")

    # Initialize client
    client = RAFTClient(host='localhost', port=2379)

    # 1. Basic key-value operations
    print("1. Basic Operations:")
    client.put('/config/app_name', 'my-app')
    value = client.get('/config/app_name')
    print(f"   GET /config/app_name = {value}")

    # 2. Atomic counter
    print("\n2. Atomic Counter:")
    for i in range(5):
        count = client.increment_counter('/counters/requests')
        print(f"   Request count: {count}")

    # 3. Compare-and-swap
    print("\n3. Compare-and-Swap:")
    client.put('/status', 'idle')
    success = client.compare_and_swap('/status', 'idle', 'processing')
    print(f"   CAS idle->processing: {success}")
    success = client.compare_and_swap('/status', 'idle', 'done')
    print(f"   CAS idle->done: {success} (expected False)")

    # 4. Lease (TTL)
    print("\n4. Lease (TTL):")
    with client.lease(10) as lease_id:
        client.put('/temp/session', 'active', lease=lease_id)
        print(f"   Created ephemeral key with 10s TTL")

    # 5. Distributed lock
    print("\n5. Distributed Lock:")
    lock = DistributedLock(client, 'my-resource')
    if lock.acquire():
        print("   Lock acquired")
        time.sleep(1)
        lock.release()
        print("   Lock released")

    # 6. Leader election
    print("\n6. Leader Election:")
    election = LeaderElection(client, 'my-service', 'node-1')
    if election.campaign():
        print(f"   Node {election.node_id} became leader")
        time.sleep(2)
        election.resign()
        print("   Resigned leadership")
    else:
        leader = election.get_leader()
        print(f"   Current leader: {leader}")

    # 7. Watch for changes
    print("\n7. Watch (run 'etcdctl put /watch/test value' in another terminal):")

    def on_change(event):
        print(f"   Change detected: {event.key.decode('utf-8')} = {event.value.decode('utf-8')}")

    watch_id = client.watch('/watch/', on_change, prefix=True)
    print("   Watching /watch/* for 5 seconds...")
    time.sleep(5)

    print("\n=== Examples Complete ===")

File: skills/distributed-systems/consensus-raft/resources/examples/go/raft_client.go
package main

/*
Production-ready etcd RAFT client in Go

Demonstrates:
- Key-value operations
- Transactions (atomic compare-and-swap)
- Leases (TTL-based keys)
- Watches (real-time notifications)
- Distributed locking
- Leader election

Requirements:
	go get go.etcd.io/etcd/client/v3

Usage:
	go run raft_client.go
*/

import (
	"context"
	"fmt"
	"log"
	"time"

	clientv3 "go.etcd.io/etcd/client/v3"
	"go.etcd.io/etcd/client/v3/concurrency"
)

// RAFTClient wraps etcd client with common operations
type RAFTClient struct {
	client *clientv3.Client
}

// NewRAFTClient creates new etcd client
func NewRAFTClient(endpoints []string) (*RAFTClient, error) {
	cli, err := clientv3.New(clientv3.Config{
		Endpoints:   endpoints,
		DialTimeout: 5 * time.Second,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to connect: %w", err)
	}

	return &RAFTClient{client: cli}, nil
}

// Close closes the client connection
func (c *RAFTClient) Close() error {
	return c.client.Close()
}

// Put writes key-value pair (goes through RAFT consensus)
func (c *RAFTClient) Put(ctx context.Context, key, value string) error {
	_, err := c.client.Put(ctx, key, value)
	if err != nil {
		return fmt.Errorf("put failed: %w", err)
	}
	return nil
}

// Get reads value for key (linearizable read from leader by default)
func (c *RAFTClient) Get(ctx context.Context, key string) (string, error) {
	resp, err := c.client.Get(ctx, key)
	if err != nil {
		return "", fmt.Errorf("get failed: %w", err)
	}

	if len(resp.Kvs) == 0 {
		return "", fmt.Errorf("key not found")
	}

	return string(resp.Kvs[0].Value), nil
}

// GetPrefix retrieves all keys with given prefix
func (c *RAFTClient) GetPrefix(ctx context.Context, prefix string) (map[string]string, error) {
	resp, err := c.client.Get(ctx, prefix, clientv3.WithPrefix())
	if err != nil {
		return nil, fmt.Errorf("get prefix failed: %w", err)
	}

	results := make(map[string]string)
	for _, kv := range resp.Kvs {
		results[string(kv.Key)] = string(kv.Value)
	}

	return results, nil
}

// Delete removes key
func (c *RAFTClient) Delete(ctx context.Context, key string) error {
	_, err := c.client.Delete(ctx, key)
	if err != nil {
		return fmt.Errorf("delete failed: %w", err)
	}
	return nil
}

// CompareAndSwap performs atomic compare-and-swap
func (c *RAFTClient) CompareAndSwap(ctx context.Context, key, expectedValue, newValue string) (bool, error) {
	txn := c.client.Txn(ctx).
		If(clientv3.Compare(clientv3.Value(key), "=", expectedValue)).
		Then(clientv3.OpPut(key, newValue)).
		Else(clientv3.OpGet(key))

	resp, err := txn.Commit()
	if err != nil {
		return false, fmt.Errorf("transaction failed: %w", err)
	}

	return resp.Succeeded, nil
}

// IncrementCounter atomically increments counter
func (c *RAFTClient) IncrementCounter(ctx context.Context, key string) (int64, error) {
	for {
		// Get current value
		currentVal, err := c.Get(ctx, key)
		if err != nil && err.Error() != "key not found" {
			return 0, err
		}

		var current int64 = 0
		if currentVal != "" {
			fmt.Sscanf(currentVal, "%d", &current)
		}

		newVal := current + 1

		// Try CAS
		success, err := c.CompareAndSwap(ctx, key, fmt.Sprintf("%d", current), fmt.Sprintf("%d", newVal))
		if err != nil {
			return 0, err
		}

		if success {
			return newVal, nil
		}

		// CAS failed, retry
		time.Sleep(time.Millisecond)
	}
}

// PutWithLease creates key with TTL
func (c *RAFTClient) PutWithLease(ctx context.Context, key, value string, ttl int64) (clientv3.LeaseID, error) {
	// Create lease
	lease, err := c.client.Grant(ctx, ttl)
	if err != nil {
		return 0, fmt.Errorf("lease grant failed: %w", err)
	}

	// Put with lease
	_, err = c.client.Put(ctx, key, value, clientv3.WithLease(lease.ID))
	if err != nil {
		return 0, fmt.Errorf("put with lease failed: %w", err)
	}

	return lease.ID, nil
}

// KeepAlive keeps lease alive
func (c *RAFTClient) KeepAlive(ctx context.Context, leaseID clientv3.LeaseID) (<-chan *clientv3.LeaseKeepAliveResponse, error) {
	return c.client.KeepAlive(ctx, leaseID)
}

// Watch watches for changes to key
func (c *RAFTClient) Watch(ctx context.Context, key string, prefix bool) clientv3.WatchChan {
	if prefix {
		return c.client.Watch(ctx, key, clientv3.WithPrefix())
	}
	return c.client.Watch(ctx, key)
}

// DistributedLock represents a distributed lock
type DistributedLock struct {
	session *concurrency.Session
	mutex   *concurrency.Mutex
}

// NewDistributedLock creates new distributed lock
func (c *RAFTClient) NewDistributedLock(ctx context.Context, lockName string, ttl int) (*DistributedLock, error) {
	session, err := concurrency.NewSession(c.client, concurrency.WithTTL(ttl))
	if err != nil {
		return nil, fmt.Errorf("session creation failed: %w", err)
	}

	mutex := concurrency.NewMutex(session, "/locks/"+lockName)

	return &DistributedLock{
		session: session,
		mutex:   mutex,
	}, nil
}

// Lock acquires the distributed lock
func (l *DistributedLock) Lock(ctx context.Context) error {
	return l.mutex.Lock(ctx)
}

// Unlock releases the distributed lock
func (l *DistributedLock) Unlock(ctx context.Context) error {
	return l.mutex.Unlock(ctx)
}

// Close closes the session
func (l *DistributedLock) Close() error {
	return l.session.Close()
}

// LeaderElection manages leader election
type LeaderElection struct {
	session  *concurrency.Session
	election *concurrency.Election
	nodeID   string
}

// NewLeaderElection creates new leader election
func (c *RAFTClient) NewLeaderElection(ctx context.Context, electionName, nodeID string, ttl int) (*LeaderElection, error) {
	session, err := concurrency.NewSession(c.client, concurrency.WithTTL(ttl))
	if err != nil {
		return nil, fmt.Errorf("session creation failed: %w", err)
	}

	election := concurrency.NewElection(session, "/elections/"+electionName)

	return &LeaderElection{
		session:  session,
		election: election,
		nodeID:   nodeID,
	}, nil
}

// Campaign attempts to become leader (blocking)
func (e *LeaderElection) Campaign(ctx context.Context) error {
	return e.election.Campaign(ctx, e.nodeID)
}

// Resign resigns from leadership
func (e *LeaderElection) Resign(ctx context.Context) error {
	return e.election.Resign(ctx)
}

// Leader returns current leader
func (e *LeaderElection) Leader(ctx context.Context) (string, error) {
	resp, err := e.election.Leader(ctx)
	if err != nil {
		return "", err
	}
	return string(resp.Kvs[0].Value), nil
}

// Close closes the session
func (e *LeaderElection) Close() error {
	return e.session.Close()
}

// Example usage
func main() {
	fmt.Println("=== RAFT Client Examples (Go) ===\n")

	// Initialize client
	client, err := NewRAFTClient([]string{"localhost:2379"})
	if err != nil {
		log.Fatalf("Failed to create client: %v", err)
	}
	defer client.Close()

	ctx := context.Background()

	// 1. Basic key-value operations
	fmt.Println("1. Basic Operations:")
	err = client.Put(ctx, "/config/app_name", "my-go-app")
	if err != nil {
		log.Printf("   PUT error: %v", err)
	}

	value, err := client.Get(ctx, "/config/app_name")
	if err != nil {
		log.Printf("   GET error: %v", err)
	} else {
		fmt.Printf("   GET /config/app_name = %s\n", value)
	}

	// 2. Atomic counter
	fmt.Println("\n2. Atomic Counter:")
	for i := 0; i < 5; i++ {
		count, err := client.IncrementCounter(ctx, "/counters/requests")
		if err != nil {
			log.Printf("   INCREMENT error: %v", err)
		} else {
			fmt.Printf("   Request count: %d\n", count)
		}
	}

	// 3. Compare-and-swap
	fmt.Println("\n3. Compare-and-Swap:")
	client.Put(ctx, "/status", "idle")
	success, _ := client.CompareAndSwap(ctx, "/status", "idle", "processing")
	fmt.Printf("   CAS idle->processing: %v\n", success)
	success, _ = client.CompareAndSwap(ctx, "/status", "idle", "done")
	fmt.Printf("   CAS idle->done: %v (expected false)\n", success)

	// 4. Lease (TTL)
	fmt.Println("\n4. Lease (TTL):")
	leaseID, err := client.PutWithLease(ctx, "/temp/session", "active", 10)
	if err != nil {
		log.Printf("   Lease error: %v", err)
	} else {
		fmt.Printf("   Created ephemeral key with 10s TTL (lease: %d)\n", leaseID)
	}

	// 5. Distributed lock
	fmt.Println("\n5. Distributed Lock:")
	lock, err := client.NewDistributedLock(ctx, "my-resource", 60)
	if err != nil {
		log.Printf("   Lock creation error: %v", err)
	} else {
		defer lock.Close()

		err = lock.Lock(ctx)
		if err != nil {
			log.Printf("   Lock acquire error: %v", err)
		} else {
			fmt.Println("   Lock acquired")
			time.Sleep(1 * time.Second)
			lock.Unlock(ctx)
			fmt.Println("   Lock released")
		}
	}

	// 6. Leader election
	fmt.Println("\n6. Leader Election:")
	election, err := client.NewLeaderElection(ctx, "my-go-service", "node-1", 60)
	if err != nil {
		log.Printf("   Election creation error: %v", err)
	} else {
		defer election.Close()

		// Try to become leader (non-blocking)
		ctx2, cancel := context.WithTimeout(ctx, 2*time.Second)
		defer cancel()

		err = election.Campaign(ctx2)
		if err == nil {
			fmt.Printf("   Node %s became leader\n", election.nodeID)
			time.Sleep(1 * time.Second)
			election.Resign(ctx)
			fmt.Println("   Resigned leadership")
		} else {
			leader, _ := election.Leader(ctx)
			fmt.Printf("   Current leader: %s\n", leader)
		}
	}

	// 7. Watch for changes
	fmt.Println("\n7. Watch (run 'etcdctl put /watch/test value' in another terminal):")
	fmt.Println("   Watching /watch/* for 5 seconds...")

	watchChan := client.Watch(ctx, "/watch/", true)

	ctx3, cancel3 := context.WithTimeout(ctx, 5*time.Second)
	defer cancel3()

	go func() {
		for watchResp := range watchChan {
			for _, event := range watchResp.Events {
				fmt.Printf("   Change detected: %s = %s\n", event.Kv.Key, event.Kv.Value)
			}
		}
	}()

	<-ctx3.Done()

	fmt.Println("\n=== Examples Complete ===")
}
